functionary/core/models/scheduled_task.py

```diff
@@ -0,0 +1,148 @@
+""" Scheduled Task model """
+import uuid
+
+import jsonschema
+from django.conf import settings
+from django.core.exceptions import ValidationError
+from django.core.serializers.json import DjangoJSONEncoder
+from django.db import models
+from django_celery_beat.models import PeriodicTask
+
+from core.models import ModelSaveHookMixin
+
+
+class ScheduledTask(ModelSaveHookMixin, models.Model):
+    """A ScheduledTask is the scheduled execution of a task
+    This model should always be queried with environment as one of the filter
+    parameters. The indices are intentionally setup this way as all requests for task
+    data happen in the context of a specific environment.
+    Attributes:
+        id: unique identifier (UUID)
+        function: the function that this task is an execution of
+        environment: the environment that this task belongs to. All queryset filtering
+                     should include an environment.
+        parameters: JSON representing the parameters that will be passed to the function
+        creator: the user that initiated the task
+        created_at: task creation timestamp
+        updated_at: task updated timestamp
+        periodic_task: The celery-beat periodic-task associated with this scheduled task
+    """
+
+    PENDING = "PENDING"
+    ACTIVE = "ACTIVE"
+    PAUSED = "PAUSED"
+    ERROR = "ERROR"
+    ARCHIVED = "ARCHIVED"
+
+    STATUS_CHOICES = [
+        (PENDING, "Pending"),
+        (ACTIVE, "Active"),
+        (PAUSED, "Paused"),
+        (ERROR, "Error"),
+        (ARCHIVED, "Archived"),
+    ]
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    name = models.CharField(max_length=200, blank=False)
+    description = models.TextField(blank=True)
+    function = models.ForeignKey(to="Function", on_delete=models.CASCADE)
+    environment = models.ForeignKey(to="Environment", on_delete=models.CASCADE)
+    parameters = models.JSONField(encoder=DjangoJSONEncoder)
+    status = models.CharField(max_length=16, choices=STATUS_CHOICES, default=PENDING)
+    creator = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.PROTECT)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+    last_run_at = models.DateTimeField(blank=True, null=True)
```

Comment:
I didn't agree with how Celery use that field. In their implementation, that field resets to `None` when the `PeriodicTask` is paused which I believe is misleading to the user.

https://django-celery-beat.readthedocs.io/en/latest/_modules/django_celery_beat/models.html#PeriodicTask

Reply:
I agree that it getting set to `None` when the PeriodicTask is disabled is a bummer.  I wouldn't personally consider it a deal breaker to just not have the `last_run_at` time available on ScheduledTasks that are paused.  But if you really want to preserve it, that's fine too.  I think you need this `last_run_at` to be set inside `run_scheduled_task` rather than where you're setting it now though.  You want the time to reflect the last time that the ScheduleTask spawned a Task, not the last time that one of its spanwed Tasks completed.  Be sure to wrap the `last_run_at` setting and the task creation in a transaction if you do that.  Alternatively, I suppose you could add the `last_run_at` setting time to the `post_create` on the Task model itself, and then you get the transaction handling for free.  I could kind of go either way on it, so I'll let you decide which way you want to handle it.